Add custom images module

ctx opened the pull request.

ctx committed:
Add custom image module

href commented on changelogs/fragments/add-costom-images-module.yml:
```diff
@@ -0,0 +1,2 @@
+major_changes:
```
The filename of this fragment has a typo (`costom`).

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+module: custom_images
+short_description: Manage custom images on the cloudscale.ch IaaS service
+description:
+  - Import, modify and delete Custom images.
```
The case should be `custom`, not `Custom`.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+    choices: [ present, absent ]
+    default: present
+    type: str
+  importstatus:
```
I would prefer `import_status`.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+  importstatus:
+    description: Get the import status of a new custom image.
+    type: bool
+    choices: [ true, false]
```
There's a space missing at `false]`.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+    name: my-custom-image
+    url: https://ubuntu.com/downloads/hirsute.img
+    slug: SLUG
+    user_data_handling: pass-throug
```
This is missing an `h`.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+  cloudscale_ch.cloud.custom_images:
+    name: my-custom-image
+    url: https://ubuntu.com/downloads/hirsute.img
+    slug: SLUG
```
I think this example could be improved:

`name: My Custom Image` (to differentiate between name and slug).
`slug: my-custom-image` (usually lower-cased and written with dashes).

Compare with the other examples as well.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+  cloudscale_ch.cloud.custom_images:
+    name: my-custom-image
+    slug: SLUG
+    user_data_handling: pass-throug
```
In most cases, `extend-cloud-config` is more likely to be what the user wants. So I would use that in these examples.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+        self._resource_data = self.init_resource()
+
+        # Set import status
+        importstatus = False
```
This whole block could also be written as follows:
```python
importstatus = self._module.params.pop('importstatus', False)
```

resmo commented on tests/integration/targets/custom_images/tasks/tests.yml:
```diff
@@ -0,0 +1,184 @@
+---
+- name: Deleta a non existent custom image by uuid
```
typo

resmo commented on tests/integration/targets/custom_images/tasks/tests.yml:
```diff
@@ -0,0 +1,184 @@
+    - '415cded5-d57c-4dc9-88fb-ececd38fb8ea'
+  ignore_errors: true
+
+- name: Deleta a non existent custom image by name
```
copy/paste typo

resmo commented on tests/integration/targets/custom_images/tasks/tests.yml:
```diff
@@ -0,0 +1,184 @@
+    state: absent
+  loop:
+    - '415cded5-d57c-4dc9-88fb-ececd38fb8ea'
+  ignore_errors: true
```
why the ignore errors? we don't expect an error here.

resmo commented on tests/integration/targets/custom_images/tasks/tests.yml:
```diff
@@ -0,0 +1,184 @@
+  cloudscale_ch.cloud.custom_images:
+    name: this-image-is-non-existent
+    state: absent
+  ignore_errors: true
```
why the ignore errors? we don't expect an error here.

resmo commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+
+DOCUMENTATION = r'''
+---
+module: custom_images
```
modules must not be named plural, exceptions are facts modules.

href commented on plugins/modules/custom_images.py:
```diff
@@ -0,0 +1,315 @@
+                key = 'name'
+            elif self._module.params['slug'] is not None:
+                key = 'slug'
+            elif self._module.params['tags'] is not None:
```
I would add an `else` clause, that raises an error if none of the other statements are True.

ctx replied:
The module requires one of the parameters `name`, `uuid`, `slug` or `tags`: `required_one_of=(('name', 'uuid', 'slug', 'tags'),)`.

The `uuid` case is handled before, so the `else` clause would be dead code.

resmo reviewed (changes requested):
looks good, except the module should be renamed to singular.

href reviewed (commented):
I have few initial comments from looking at the source code. If you don't mind I'll hand this back to you, you fix these issues and then let me know. Then I'll do a functional QA.

href commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,312 @@
+    name: "My Custom Image"
+    url: https://ubuntu.com/downloads/hirsute.img
+    slug: my-custom-image
+    user_data_handling: extended-cloud-config
```
This should read `extend-cloud-config`.

href commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,312 @@
+    description:
+      - The file format of the image referenced in the url. Currently only raw
+        is supported.
+    default: "raw"
```
We shouldn't use a default here, as the API forces you to write this out. Not that that's necessarily good ergonomics, but I would stay true to the API here.

href commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,314 @@
+- name: Update custom image
+  cloudscale_ch.cloud.custom_image:
+    name: "My Custom Image"
+    slug: SLUG
```
I would lower-case this, use `my-custom-image` or something else (as mentioned elsewhere before).

href commented:
I found two issues that need some attention:

1. When I specify the following resource on an otherwise empty account, I get an `HTTP 405: Method Not Allowed Error`. I can image this happens because I don't pass a URL, but I would expect a more descriptive error:

```yaml
- name: Upload custom alpine image
  cloudscale_ch.cloud.custom_image:
    name: Custom Alpine Image
    slug: alpine
    source_format: raw
    user_data_handling: extend-cloud-config
    zones:
      - lpg1
      - rma1
```

2. When I import the image, the return value is different in structure, than if I query the image. I can imagine why this is the case, but as a user I find this surprising. For example, according to the docs, the return value should include a slug. However, this is not the case when I run the following playbook. Only when running it again does it work (in which case the image has already been imported):

```yaml
#!/usr/bin/env ansible-playbook

- name: Start Alpine Server
  hosts: localhost
  tasks:

    - name: Upload custom alpine image
      cloudscale_ch.cloud.custom_image:
        name: Custom Alpine Image 2
        url: https://at-images.objects.lpg.cloudscale.ch/alpine
        slug: alpine
        source_format: raw
        user_data_handling: extend-cloud-config
        zones:
          - lpg1
          - rma1
      register: custom_image

    - name: Wait until import succeeded
      cloudscale_ch.cloud.custom_image:
        uuid: '{{ custom_image.uuid }}'
        import_status: true
      retries: 15
      delay: 5
      register: import_status
      until: import_status.status == 'success'

    - name: Launch a server with a custom image
      cloudscale_ch.cloud.server:
        name: alpine-test
        image: 'custom:{{ custom_image.slug }}'
        flavor: flex-2
        zone: lpg1
```

resmo commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,314 @@
+                    if self._module.params[key] is not None:
+                        if self.find_difference(key,
+                                                resource,
+                                                self._module.params[key]):
```
btw ansible's max line limit policy is 160, just saying

href replied:
I generally nag people to use 79, and @ctx has had to pass a few Python reviews of mine already 😉

ctx committed:
Return all values and wait until an import started

Return values from both urls, with and without import
Wait until one can do a GET request for a newly imported image

href commented on plugins/modules/custom_image.py:
```diff
@@ -178,22 +172,101 @@
+        if api_call == 'custom-images':
+            api_call = 'custom-images/import'
+        else:
+            api_call = re.sub(r"/", "/import/", api_call)
```
Why do you use a regex for this? Wouldn't it be simpler to just append `import`?

href commented on plugins/modules/custom_image.py:
```diff
@@ -178,22 +172,101 @@
+            url = '%s%s/%s' % (self._api_url, api_call, ret['uuid'])
+            info = {}
+            info['status'] = 0
+            while info['status'] != 200:
```
I think this should be limited to a certain amount of time. For example:

```python
for _ in range(10):
    ....
```

As it stands it would try this forever if something goes wrong, which is probably not what we want.

href reviewed (changes requested):
From a user perspective I think this is a great improvement. I like being able to upload and to immediately launch a server with my new image, with a nice retry status as it waits for the import to finish.

Aside from minor code changes (and some still not addressed feedback from the last review) I noticed two things:

- `status`: I think this should be named `import_status` as before. Otherwise we have `state` and `status`, which I find a bit too close in name.
- The `status` or `import_status` is no longer documented.

My updated example that I'm using is here, with how I wish it would look like (`import_status`):

```yaml
- name: Upload custom alpine image
  cloudscale_ch.cloud.custom_image:
    name: Custom Alpine Image
    url: https://at-images.objects.lpg.cloudscale.ch/alpine
    slug: alpine
    source_format: raw
    user_data_handling: extend-cloud-config
    zones:
      - lpg1
      - rma1
  register: custom_image
  retries: 15
  delay: 5
  until: custom_image.import_status == 'success'
```

I think something like this, together with how to start a server in a next step, would also make a good example that should be incorporated somewhere. In most cases that's probably what users would want to do.

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,439 @@
+    url: https://ubuntu.com/downloads/hirsute.img
+    slug: my-custom-image
+    user_data_handling: extend-cloud-config
+    zones: LPG1
```
This should be lowercase. Not sure if the API even allows this, but it's certainly not how it's usually written.

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,439 @@
+    if module.params['state'] == "absent":
+        result = cloudscale_custom_image.absent()
+    else:
+        cloudscale_custom_image.resource_name = 'custom-images'
```
Why do you have to override this? AFAICS it's already set to this value when instantiating the object above.

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,439 @@
+
+        if not response and response_import:
+            ensure_keys = {
+                'created_at': '2050-01-01T00:00:00.000000Z',
```
What's the reason you set this to an arbitrary date somewhat far in the future. Is this so you can compare dates?
I would prefer to use `'null'` or `'None'` or even a Python `None` if possible instead of this magic value.

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,439 @@
+        if api_call == 'custom-images':
+            api_call = 'custom-images/import'
+        else:
+            api_call = api_call.replace('/', '/import/')
```
Maybe add a comment that this is to transform URLs including an image UUID would be good. Otherwise this is quite opaque.

gaudenz reviewed (changes requested):
As discussed we agreed to not overwrite the query function from `AnsibleCloudscaleBase` and to only overwrite the `_get` function to fetch resources from `custom-images` and `custom-images/import`.

Custom images should be identified with either their UUID or their name like all other resources. Listing custom images by slug or tag is probably better done using the uri module or in a separate `custom_image_facts` module.

ctx committed:
Make name unique and don't list images by slug/tag

Overwrite _post/_get/_post_or_patch and use the default query function.

resmo commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,441 @@
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_one_of=(('name', 'uuid'),),
+        required_if=(('state', 'absent', ('uuid',)),),
```
It should be possible to delete by name as long as only one image is found with that name.

ctx committed:
Allow to delete an image by name

ctx committed:
Don't fail when a failed import exists

Import a new image when a failed import with that name exists
Add tests with bad url
Fail examples and tests when the url is not reachable

resmo commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,479 @@
+      - How user_data will be handled when creating a server. There are
+        currently two options, "pass-through" and "extend-cloud-config".
+    type: str
+    choices: [ pass-through , extend-cloud-config ]
```
```diff
-     choices: [ pass-through , extend-cloud-config ]
+     choices: [ pass-through, extend-cloud-config ]
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+#    uuid: "{{ item }}"
+#    state: absent
+#  loop:
+#    - '415caea5-da7c-4aaa-aaaa-ececd38fb8ea'
```
If you are just testing one UUID, there is no need to have a loop:
```suggestion
#- name: Delete a non existent custom image by uuid
#  cloudscale_ch.cloud.custom_image:
#    uuid: '415caea5-da7c-4aaa-aaaa-ececd38fb8ea'
#    state: absent
```
And the assert part is missing from this (commented) test.

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+- name: Verify Fail delete a non existend custom image by name
+  assert:
+    that:
+      - delete is not changed
```
Shouldn't this directly test for failure?
```suggestion
      - delete is failed
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+- name: Verify Fail import a custom image without url
+  assert:
+    that:
+      - failed_import.msg.startswith('Cannot import a new image without url.')
```
This should also verify that the import indeed fails.
```suggestion
      - failed_import is failed
      - failed_import.msg.startswith('Cannot import a new image without url.')
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+- name: Verify Fail import a custom image with bad url
+  assert:
+    that:
+      - failed_import.error_message.startswith('Expected HTTP 200, got HTTP 40')
```
Same as above:
```suggestion
      - failed_import is failed
      - failed_import.error_message.startswith('Expected HTTP 200, got HTTP 40')
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+      - import_status is not changed
+      - import_status.name == "{{ cloudscale_resource_prefix }}-test-image2"
+
+- name: Get newest image by name
```
As managing more than one image with the same name is no longer supported:
```suggestion
- name: Get image by name
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+  cloudscale_ch.cloud.custom_image:
+    name: "{{ cloudscale_resource_prefix }}-test-image2"
+  register: image_by_name
+- name: Verify get newest image by name
```
```suggestion
- name: Verify get image by name
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+- name: Verify update slug of a custom image
+  assert:
+    that:
+      - image is changed
```
I would also check that the slug name is correct:
```suggestion
      - image is changed
      - image.slug == 'ansible-image-slug'
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+    slug: ansible-image-slug
+  register: image
+  loop:
+    - "{{ image2.uuid }}"
```
No need to use a loop. Especially as the assert below does not use a loop:
```suggestion
- name: Update slug of a custom image
  cloudscale_ch.cloud.custom_image:
    uuid: "{{ image2.uuid }}"
    slug: ansible-image-slug
  register: image
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+  cloudscale_ch.cloud.custom_image:
+    uuid: "{{ item }}"
+  loop:
+    - "{{ image2.uuid }}"
```
```suggestion
- name: Get custom image with updated slug
  cloudscale_ch.cloud.custom_image:
    uuid: "{{ image2.uuid }}"
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+      project: luna
+  register: image
+  loop:
+    - "{{ image2.uuid }}"
```
```suggestion
- name: Update tags of a custom image
  cloudscale_ch.cloud.custom_image:
    uuid: "{{ image2.uuid }}"
    tags:
      project: luna
  register: image
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+  assert:
+    that:
+      - image is changed
+      - image.results[0].tags == "project: luna"
```
```suggestion
      - image.tags == "project: luna"
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+    user_data_handling: 'extend-cloud-config'
+  register: image
+  loop:
+    - "{{ image2.uuid }}"
```
```suggestion
- name: Update user_data_handling of a custom image
  cloudscale_ch.cloud.custom_image:
    uuid: "{{ image2.uuid }}"
    user_data_handling: 'extend-cloud-config'
  register: image
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+- name: Verify update user_data_handling of a custom image
+  assert:
+    that:
+      - image is changed
```
Also check that the changed parameter indeed has the correct value:
```suggestion
      - image is changed
      - image.user_data_handling = 'extend-cloud-config'
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+    uuid: "{{ item }}"
+  loop:
+    - "{{ image2.uuid }}"
+  register: image
```
```suggestion
- name: Get custom image with updated user_data_handling
  cloudscale_ch.cloud.custom_image:
    uuid: "{{ image2.uuid }}"
  register: image
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+      - image is changed
+      - image.results[0].slug == "yet-another-slug"
+      - image.results[0].tags == "project: jupiter"
+      - image.results[0].user_data_handling == "pass-through"
```
```suggestion
- name: Update slug, tags and user_data_handling of a custom image
  cloudscale_ch.cloud.custom_image:
    uuid: "{{ image2.uuid }}"
    slug: yet-another-slug
    tags:
      project: jupiter
    user_data_handling: 'pass-through'
  register: image
- name: Verify update slug, tags and user_data_handling of a custom image
  assert:
    that:
      - image is changed
      - image.slug == "yet-another-slug"
      - image.tags == "project: jupiter"
      - image.user_data_handling == "pass-through"
```

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+  register: image_list
+  until: image_list is not failed
+  retries: 5
+  delay: 3
```
Why do we have to retry this? I think this should not fail randomly. If it does it should be fixed ;-).

gaudenz commented on tests/integration/targets/custom_image/tasks/tests.yml:
```diff
@@ -0,0 +1,267 @@
+  assert:
+    that:
+      - image1 is changed
+      - image1.state == 'absent'
```
Nitpick: Please insert a blank line here.
```suggestion
      - image1.state == 'absent'
      
```

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,479 @@
+  uri:
+    url: 'https://api.cloudscale.ch/v1/custom-images'
+    headers:
+      Authorization: 'Bearer {{ cloudscale_api_token }}'
```
While this works, I suggest to use the canonical way of providing the token for the Ansible modules:
```suggestion
      Authorization: 'Bearer {{ query("env", "CLOUDSCALE_API_TOKEN") }}'
```

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,479 @@
+
+from copy import deepcopy
+
+from time import sleep
```
This is not used anywhere (discovered using pyflakes).

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,479 @@
+        state=dict(type='str', default='present',
+                   choices=('present', 'absent')),
+        zones=dict(type='list', elements='str'),
+        source_format=dict(type='str', choices=('raw', )),
```
Like we are not restricting the list of possible zones I would also not restrict the list of possible source formats. Without this restriction the module should automatically work with other source formats as they become supported by the API.
```suggestion
        source_format=dict(type='str'),
```

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,479 @@
+        if self._module.params['url']:
+            return self._post_or_patch("%s" % api_call, 'POST', data)
+        else:
+            self._module.fail_json(msg="Cannot import a new image without url.")
```
If I understand this right, this is only to provide a customized error message instead of the error message provided by the API in case the `url` parameter is missing on image creation. I'm not sure this is even necessary, but if you think it is, I would prefer to overwrite the `create` method with something like this:
```
def create(self, resource, data=None):
    if not self._module.params['url']:
        self._module.fail_json(msg="Cannot import a new image without url.")
    return super(AnsibleCloudscaleCustomImage, self).create(resource, data)
```

gaudenz commented on plugins/modules/custom_image.py:
```diff
@@ -0,0 +1,479 @@
+
+    def check_for_uuid(self, resources):
+        if 'uuid' in self._module.params and self._module.params['uuid']:
+            res = []
```
Why is the default return value if no resource with the UUID is found an empty list. Shouldn't this be `None` (or maybe an empty dict)?

gaudenz reviewed (changes requested):
As discussed I'm sending you the first part of the review which mostly covers the tests. Two additional remarks:
* A test for the correct failure if two custom images with the same name already exist in an account. For this you probably have to create the second image with the same name using the`uri` module.
* A test to verify that it's possible to change the name of an image if an UUID parameter is also present.
IIRC there are examples for both these tests in the tests for the server module.

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Update tests/integration/targets/custom_image/tasks/tests.yml

Co-authored-by: Gaudenz Steinlin <gaudenz@users.noreply.github.com>

ctx committed:
Add more tests and fix old ones

ctx committed:
Add all suggestions by reviewers

ctx committed:
Fix flake and yamllint warnings

ctx committed:
small fixes

ctx committed:
fix doc
